experimental/parser/math.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"math/big"
+	"math/bits"
+)
+
+type parseIntResult struct {
+	small        uint64
+	big          *big.Int
+	hasThousands bool
+}
+
+// parseInt parses an integer into a uint64 or, on overflow, into a big.Int.
+//
+// This function ignores any thousands separator underscores in digits.
+func parseInt(digits string, base byte) (result parseIntResult, ok bool) {
+	for _, r := range digits {
+		if r == '_' {
+			result.hasThousands = true
+			continue
+		}
+
+		digit := parseDigit(r)
+		if digit >= base {
+			return result, false
+		}
+
+		if result.big != nil {
+			result.big.Mul(result.big, big.NewInt(int64(base)))
```

Comment:
We should create the base representation only once and memoize in a local variable, instead of re-creating for every digit.